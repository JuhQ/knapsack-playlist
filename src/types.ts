export interface YoutubeItem {
  id: string;
  title: string;
  seconds: number;
  rating: number;
}
