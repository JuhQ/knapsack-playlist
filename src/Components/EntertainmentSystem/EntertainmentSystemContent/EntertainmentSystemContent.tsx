import React from "react";
import { Grid } from "semantic-ui-react";

import { YoutubeItem } from "../../../types";
import { sumSeconds } from "../../../Utils/math";
import Player from "../../Player/Player";
import Playlist from "../../Playlist/Playlist";

interface Props {
  list: YoutubeItem[];
  playlist: YoutubeItem[];
  currentSong: YoutubeItem;
  playlistLength: number;
  length: number;
  diff: number;
  onEnd: () => void;
  onChange: (item: YoutubeItem) => void;
}

const EntertainmentSystemContent: React.FC<Props> = ({
  list,
  playlist,
  currentSong,
  playlistLength,
  length,
  diff,
  onEnd,
  onChange,
}: Props) => (
  <>
    <Grid verticalAlign="bottom">
      <Grid.Column computer={10} mobile={16} textAlign="right">
        <h2>{currentSong.title}</h2>
      </Grid.Column>
      <Grid.Column computer={5} mobile={16} textAlign="center">
        <span>Playlist length: {playlistLength} seconds</span>
        <br />
        <span>Tried to generate {length} seconds</span>
        <br />
        {diff === 0 && (
          <strong>
            Great success! Playlist is exactly as long as we wanted!
            <span role="img" aria-label="Success">
              ðŸ‘Œ
            </span>
            <br />
          </strong>
        )}
        {diff > 0 && (
          <>
            Difference: {diff} seconds
            <br />
          </>
        )}
        <span>Queue length: {sumSeconds(playlist)} seconds</span>
        <br />
        <span>Playlist size: {list.length} songs</span>
        <br />
        <span>Queue size: {playlist.length} songs</span>
      </Grid.Column>
    </Grid>
    <Grid>
      <Grid.Column computer={10} mobile={16} textAlign="right">
        <Player id={currentSong.id} onEnd={onEnd} />
      </Grid.Column>
      <Grid.Column computer={5} mobile={16}>
        <Playlist list={playlist} onChange={onChange} />
      </Grid.Column>
    </Grid>
  </>
);

export default EntertainmentSystemContent;
