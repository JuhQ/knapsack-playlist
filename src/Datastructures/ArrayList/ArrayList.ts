/**
 * ArrayList implementation in typescript
 */
class ArrayList<T> {
  private list: T[];

  private index: number;

  constructor(size: number) {
    this.list = Array<T>(size);
    this.index = 0;
  }

  // Time complexity O(1)
  private isFull(): boolean {
    return this.index + 1 === this.list.length;
  }

  // Time complexity O(n)
  // Space complexity O(n)
  private grow(): T[] {
    // reserve three times as much space for the next list, so we don't need to do this operation too often
    const list = new Array<T>(this.size() * 3);

    for (let i = 0; i < this.size(); i++) {
      list[i] = this.list[i];
    }

    return list;
  }

  // Time complexity O(1), worst case O(n)
  push(value: T): void {
    if (this.isFull()) {
      this.list = this.grow();
    }

    this.list[this.index] = value;
    this.index++;
  }

  // Time complexity O(n)
  // Space complexity O(n)
  slice(start?: number, end?: number): ArrayList<T> {
    const sliced = this.getAsArray().slice(start, end);

    const list = new ArrayList<T>(sliced.length);

    for (let i = 0; i < sliced.length; i++) {
      list.push(sliced[i]);
    }

    return list;
  }

  // remove item from the beginning of the queue and return that value
  // Time complexity O(n)
  shift(): T | undefined {
    this.index--;
    return this.list.shift();
  }

  // Time complexity O(n)
  // Space complexity O(n)
  map<K>(fn: (value: T) => K): ArrayList<K> {
    const list = this.getAsArray();
    const newList = new ArrayList<K>(this.size());

    for (let i = 0; i < list.length; i++) {
      newList.push(fn(list[i]));
    }

    return newList;
  }

  // Time complexity O(n)
  // Space complexity O(n)
  filter(fn: (value: T) => boolean): ArrayList<T> {
    const list = this.getAsArray();
    const newList = new ArrayList<T>(1);

    for (let i = 0; i < list.length; i++) {
      if (fn(list[i])) {
        newList.push(list[i]);
      }
    }

    return newList;
  }

  // Time complexity O(n)
  // Space complexity O(n)
  find<K>(fn: (value: T) => K): T | undefined {
    const list = this.getAsArray();

    for (let i = 0; i < list.length; i++) {
      if (fn(list[i])) {
        return list[i];
      }
    }

    return undefined;
  }

  // TODO: implement proper sort algorithm
  sort(fn: (a: T, b: T) => number): ArrayList<T> {
    const sorted = [...this.list].sort(fn);

    const list = new ArrayList<T>(this.size());
    for (let i = 0; i < this.size(); i++) {
      list.push(sorted[i]);
    }

    return list;
  }

  // Time complexity O(1)
  size(): number {
    return this.index;
  }

  // Time complexity O(1)
  at(index: number): T {
    return this.list[index];
  }

  /**
   * Return all values from the list, except undefined values
   * This will cause issue if given type would be undefined, but in this project undefined lists will never be used
   * Undefined is removed because the grow method will reserve extra space
   * Time complexity O(n)
   * Space complexity O(n)
   */
  getAsArray(): T[] {
    const list: T[] = [];
    for (let i = 0; i < this.list.length; i++) {
      if (this.list[i] !== undefined) {
        list.push(this.list[i]);
      }
    }

    return list;
  }
}

export default ArrayList;
