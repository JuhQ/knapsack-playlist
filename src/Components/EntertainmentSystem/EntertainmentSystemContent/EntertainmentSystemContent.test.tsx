import { shallow } from "enzyme";
import { shallowToJson } from "enzyme-to-json";
import React from "react";
import { act, create } from "react-test-renderer";

import ArrayList from "../../../Datastructures/ArrayList/ArrayList";
import Queue from "../../../Datastructures/queue/queue";
import { YoutubeItem } from "../../../types";
import EntertainmentSystemContent from "./EntertainmentSystemContent";

describe("EntertainmentSystem", () => {
  it("should render", () => {
    const list = new ArrayList<YoutubeItem>(2);
    list.push({ id: "1", title: "test", seconds: 60, rating: 1 });
    list.push({ id: "2", title: "test 2", seconds: 60, rating: 1 });
    const list2 = new ArrayList<YoutubeItem>(1);
    list2.push({ id: "2", title: "test 2", seconds: 60, rating: 1 });
    const queue = new Queue();
    queue.merge(list);
    const playlist = new Queue();
    playlist.merge(list2);

    const component = create(
      <EntertainmentSystemContent
        list={queue}
        playlist={playlist}
        currentSong={{ id: "1", title: "test", seconds: 60, rating: 1 }}
        length={120}
        diff={0}
        playlistLength={120}
        onEnd={() => null}
        onChange={() => null}
      />
    );
    const tree = component.toJSON();
    expect(tree).toMatchSnapshot();
  });

  it("should render diff", () => {
    const list = new ArrayList<YoutubeItem>(2);
    list.push({ id: "1", title: "test", seconds: 55, rating: 1 });
    list.push({ id: "2", title: "test 2", seconds: 60, rating: 1 });
    const list2 = new ArrayList<YoutubeItem>(1);
    list2.push({ id: "2", title: "test 2", seconds: 60, rating: 1 });
    const queue = new Queue();
    queue.merge(list);
    const playlist = new Queue();
    playlist.merge(list2);
    const component = create(
      <EntertainmentSystemContent
        list={queue}
        playlist={playlist}
        currentSong={{ id: "1", title: "test", seconds: 55, rating: 1 }}
        length={120}
        diff={5}
        playlistLength={115}
        onEnd={() => null}
        onChange={() => null}
      />
    );
    const tree = component.toJSON();
    expect(tree).toMatchSnapshot();
  });

  it("should call song changing function on player when current song ends", () => {
    const list = new ArrayList<YoutubeItem>(2);
    list.push({ id: "1", title: "test", seconds: 55, rating: 1 });
    list.push({ id: "2", title: "test 2", seconds: 60, rating: 1 });
    const list2 = new ArrayList<YoutubeItem>(1);
    list2.push({ id: "2", title: "test 2", seconds: 60, rating: 1 });
    const queue = new Queue();
    queue.merge(list);
    const playlist = new Queue();
    playlist.merge(list2);
    const handleChange = jest.fn();
    const wrapper = shallow(
      <EntertainmentSystemContent
        list={queue}
        playlist={playlist}
        currentSong={{ id: "1", title: "test", seconds: 55, rating: 1 }}
        length={120}
        diff={5}
        playlistLength={115}
        onEnd={() => null}
        onChange={handleChange}
      />
    );

    wrapper.find("Playlist").simulate("change");
    expect(handleChange).toHaveBeenCalled();
  });

  it("should call end function on player when current song ends", () => {
    const list = new ArrayList<YoutubeItem>(2);
    list.push({ id: "1", title: "test", seconds: 55, rating: 1 });
    list.push({ id: "2", title: "test 2", seconds: 60, rating: 1 });
    const list2 = new ArrayList<YoutubeItem>(1);
    list2.push({ id: "2", title: "test 2", seconds: 60, rating: 1 });
    const queue = new Queue();
    queue.merge(list);
    const playlist = new Queue();
    playlist.merge(list2);
    const handleEnd = jest.fn();
    const wrapper = shallow(
      <EntertainmentSystemContent
        list={queue}
        playlist={playlist}
        currentSong={{ id: "1", title: "test", seconds: 55, rating: 1 }}
        length={120}
        diff={5}
        playlistLength={115}
        onEnd={handleEnd}
        onChange={() => null}
      />
    );

    wrapper.find("Player").simulate("end");
    expect(handleEnd).toHaveBeenCalled();
  });

  it("should call song changing function on playlist when changing song in playlist", () => {
    const list = new ArrayList<YoutubeItem>(2);
    list.push({ id: "1", title: "test", seconds: 55, rating: 1 });
    list.push({ id: "2", title: "test 2", seconds: 60, rating: 1 });
    const list2 = new ArrayList<YoutubeItem>(2);
    list2.push({ id: "2", title: "test 2", seconds: 60, rating: 1 });
    const queue = new Queue();
    queue.merge(list);
    const playlist = new Queue();
    playlist.merge(list2);
    const handleChange = jest.fn();
    const wrapper = shallow(
      <EntertainmentSystemContent
        list={queue}
        playlist={playlist}
        currentSong={{ id: "1", title: "test", seconds: 55, rating: 1 }}
        length={120}
        diff={5}
        playlistLength={115}
        onEnd={() => null}
        onChange={handleChange}
      />
    );

    wrapper
      .find("Playlist")
      .simulate("change", { id: "2", title: "test 2", seconds: 60, rating: 1 });

    expect(handleChange).toHaveBeenCalled();
  });

  it("should render playlist ended message once all the songs have been played", () => {
    const list = new ArrayList<YoutubeItem>(2);
    list.push({ id: "1", title: "test", seconds: 55, rating: 1 });
    list.push({ id: "2", title: "test 2", seconds: 60, rating: 1 });
    const list2 = new ArrayList<YoutubeItem>(1);
    list2.push({ id: "2", title: "test 2", seconds: 60, rating: 1 });
    const queue = new Queue();
    queue.merge(list);
    const playlist = new Queue();
    playlist.merge(list2);
    const wrapper = shallow(
      <EntertainmentSystemContent
        list={queue}
        playlist={playlist}
        currentSong={{ id: "1", title: "test", seconds: 55, rating: 1 }}
        length={120}
        diff={5}
        playlistLength={115}
        onEnd={() => null}
        onChange={() => null}
      />
    );

    wrapper.find("Player").simulate("end");
    expect(shallowToJson(wrapper)).toMatchSnapshot();
    act(() => {
      expect(shallowToJson(wrapper)).toMatchSnapshot();
    });
  });
});
