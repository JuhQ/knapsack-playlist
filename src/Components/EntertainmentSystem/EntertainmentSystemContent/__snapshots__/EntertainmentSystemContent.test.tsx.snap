// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`EntertainmentSystem should render 1`] = `
Array [
  <div
    className="ui bottom aligned grid"
  >
    <div
      className="right aligned ten wide computer sixteen wide mobile column"
    >
      <h2>
        test
      </h2>
    </div>
    <div
      className="center aligned five wide computer sixteen wide mobile column"
    >
      <span>
        Playlist duration: 
        120
         seconds
      </span>
      <br />
      <span>
        Tried to generate 
        120
         seconds
      </span>
      <br />
      <strong>
        Great success! Playlist is exactly as long as you wanted!
        <span
          aria-label="Success"
          role="img"
        >
          ðŸ‘Œ
        </span>
        <span
          aria-label="Success"
          role="img"
        >
          ðŸ¤˜
        </span>
        <br />
      </strong>
      <span>
        Queue duration: 
        60
         seconds
      </span>
      <br />
      <span>
        Playlist size: 
        2
         songs
      </span>
      <br />
      <span>
        Queue size: 
        1
         songs
      </span>
    </div>
  </div>,
  <div
    className="ui grid"
  >
    <div
      className="right aligned ten wide computer sixteen wide mobile column"
    >
      <div
        className=""
      >
        <div
          className={null}
          id={null}
        />
      </div>
    </div>
    <div
      className="five wide computer sixteen wide mobile column"
    >
      <div
        className="playlist"
      >
        <div
          className="ui items"
        >
          <a
            className="item"
            onClick={[Function]}
          >
            <div
              className="image"
            >
              <img
                src="https://i.ytimg.com/vi/2/hqdefault.jpg"
              />
            </div>
            <div
              className="content"
            >
              <div
                className="header"
              >
                test 2
              </div>
              <div
                className="meta"
              >
                1 minute
                 
                
              </div>
            </div>
          </a>
        </div>
      </div>
    </div>
  </div>,
]
`;

exports[`EntertainmentSystem should render diff 1`] = `
Array [
  <div
    className="ui bottom aligned grid"
  >
    <div
      className="right aligned ten wide computer sixteen wide mobile column"
    >
      <h2>
        test
      </h2>
    </div>
    <div
      className="center aligned five wide computer sixteen wide mobile column"
    >
      <span>
        Playlist duration: 
        115
         seconds
      </span>
      <br />
      <span>
        Tried to generate 
        120
         seconds
      </span>
      <br />
      Difference: 
      5
       seconds
      <br />
      <span>
        Queue duration: 
        60
         seconds
      </span>
      <br />
      <span>
        Playlist size: 
        2
         songs
      </span>
      <br />
      <span>
        Queue size: 
        1
         songs
      </span>
    </div>
  </div>,
  <div
    className="ui grid"
  >
    <div
      className="right aligned ten wide computer sixteen wide mobile column"
    >
      <div
        className=""
      >
        <div
          className={null}
          id={null}
        />
      </div>
    </div>
    <div
      className="five wide computer sixteen wide mobile column"
    >
      <div
        className="playlist"
      >
        <div
          className="ui items"
        >
          <a
            className="item"
            onClick={[Function]}
          >
            <div
              className="image"
            >
              <img
                src="https://i.ytimg.com/vi/2/hqdefault.jpg"
              />
            </div>
            <div
              className="content"
            >
              <div
                className="header"
              >
                test 2
              </div>
              <div
                className="meta"
              >
                1 minute
                 
                
              </div>
            </div>
          </a>
        </div>
      </div>
    </div>
  </div>,
]
`;

exports[`EntertainmentSystem should render playlist ended message once all the songs have been played 1`] = `
<Fragment>
  <Grid
    verticalAlign="bottom"
  >
    <GridColumn
      computer={10}
      mobile={16}
      textAlign="right"
    >
      <h2>
        test
      </h2>
    </GridColumn>
    <GridColumn
      computer={5}
      mobile={16}
      textAlign="center"
    >
      <span>
        Playlist duration: 
        115
         seconds
      </span>
      <br />
      <span>
        Tried to generate 
        120
         seconds
      </span>
      <br />
      Difference: 
      5
       seconds
      <br />
      <span>
        Queue duration: 
        60
         seconds
      </span>
      <br />
      <span>
        Playlist size: 
        2
         songs
      </span>
      <br />
      <span>
        Queue size: 
        1
         songs
      </span>
    </GridColumn>
  </Grid>
  <Grid>
    <GridColumn
      computer={10}
      mobile={16}
      textAlign="right"
    >
      <Player
        id="1"
        onEnd={[Function]}
      />
    </GridColumn>
    <GridColumn
      computer={5}
      mobile={16}
    >
      <Playlist
        list={
          Queue {
            "enqueue": [Function],
            "list": ArrayList {
              "index": 1,
              "list": Array [
                Object {
                  "id": "2",
                  "rating": 1,
                  "seconds": 60,
                  "title": "test 2",
                },
              ],
            },
          }
        }
        onChange={[Function]}
      />
    </GridColumn>
  </Grid>
</Fragment>
`;

exports[`EntertainmentSystem should render playlist ended message once all the songs have been played 2`] = `
<Fragment>
  <Grid
    verticalAlign="bottom"
  >
    <GridColumn
      computer={10}
      mobile={16}
      textAlign="right"
    >
      <h2>
        test
      </h2>
    </GridColumn>
    <GridColumn
      computer={5}
      mobile={16}
      textAlign="center"
    >
      <span>
        Playlist duration: 
        115
         seconds
      </span>
      <br />
      <span>
        Tried to generate 
        120
         seconds
      </span>
      <br />
      Difference: 
      5
       seconds
      <br />
      <span>
        Queue duration: 
        60
         seconds
      </span>
      <br />
      <span>
        Playlist size: 
        2
         songs
      </span>
      <br />
      <span>
        Queue size: 
        1
         songs
      </span>
    </GridColumn>
  </Grid>
  <Grid>
    <GridColumn
      computer={10}
      mobile={16}
      textAlign="right"
    >
      <Player
        id="1"
        onEnd={[Function]}
      />
    </GridColumn>
    <GridColumn
      computer={5}
      mobile={16}
    >
      <Playlist
        list={
          Queue {
            "enqueue": [Function],
            "list": ArrayList {
              "index": 1,
              "list": Array [
                Object {
                  "id": "2",
                  "rating": 1,
                  "seconds": 60,
                  "title": "test 2",
                },
              ],
            },
          }
        }
        onChange={[Function]}
      />
    </GridColumn>
  </Grid>
</Fragment>
`;
